tempesta_fw/sock.c

```diff
@@ -483,7 +551,7 @@ __ss_close(struct sock *sk, int flags)
 	ss_do_close(sk);
 	bh_unlock_sock(sk);
 	if (flags & SS_F_CONN_CLOSE)
-		SS_CALL(connection_drop, sk);
+		SS_CALL_GUARD_EXIT(connection_drop, sk);
```

Comment:
Hm, it seems I didn't get you.

`connection_drop()` is called on intentional, our side, closing only (`ss_tx_action()` and `__ss_close()`). `ss_droplink()` is called on unintentional, other side in this case, closing. In any case `ss_do_close()` is called under socket lock. So `connection_drop()` and `connection_error()` callbacks will be called concurrently in your case. Basically, client and server functions registered on the SS hooks are designed to be run in concurrently: we unlink TfwConnection from `sk->sk_user_data` before destroying it and carefully track TfwConnection reference counter to survive concurrent TfwConnection put.

About which issue exactly are you talking about? Do you see exact code places with some weakness?